<div style="color: #095AAD; font-weight: bold; font-size: 16px;">
   
# Data Science Salary - Data Cleaning </div>

This notebook focuses on cleaning and preparing the Data Science salary dataset for analysis. I will ensure data quality, handle missing values, identify and address outliers, validate data consistency, and prepare a clean dataset for comprehensive salary analysis.

**Data source**: [Kaggle - Data Science Job Salaries](https://www.kaggle.com/datasets/adilshamim8/salaries-for-data-science-jobs)

<div style="color: #095AAD; font-weight: bold; font-size: 16px;">
   
## Dataset Description</div>

The dataset contains salary information for data science professionals with the following structure:

| **Column Name** | **Description** |
|-------------|-------------|
| `work_year` | Year of employment (2020-2025) |
| `experience_level` | Experience level (EN/MI/SE/EX) |
| `employment_type` | Type of employment (FT/PT/CT/FL) |
| `job_title` | Specific job title in data science field |
| `salary` | Salary in original currency |
| `salary_currency` | Currency of salary |
| `salary_in_usd` | Salary converted to USD |
| `employee_residence` | Country of employee residence |
| `remote_ratio` | Percentage of remote work (0/50/100) |
| `company_location` | Country where company is located |
| `company_size` | Size of company (S/M/L) |

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Importing required libraries</div>

In [1]:
import pandas as pd
import numpy as np
import warnings

warnings.filterwarnings('ignore')

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Data loading</div>

In [2]:
df = pd.read_csv('salaries.csv')

print('Dataset shape:', df.shape)
print('\nFirst 5 rows:')
df.head()

Dataset shape: (136757, 11)

First 5 rows:


work_year experience_level employment_type                  job_title  \
0       2025               MI              FT             Data Scientist   
1       2025               MI              FT             Data Scientist   
2       2025               SE              FT       Data Product Manager   
3       2025               SE              FT       Data Product Manager   
4       2025               SE              FT  Machine Learning Engineer   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0  132600             USD         132600                 US           100   
1  102000             USD         102000                 US           100   
2  260520             USD         260520                 US             0   
3  140280             USD         140280                 US             0   
4  215000             USD         215000                 US             0   

  company_location company_size  
0               US            M  
1               US            M  
2               US            M  
3               US            M  
4               US            M

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Duplicate removal</div>

Checking for and removing duplicate records to ensure data quality and avoid bias in our analysis.

In [3]:
# Analyze duplicates
print('Duplicate analysis:')
print(f'Total records: {len(df)}')
print(f'Duplicates found: {df.duplicated().sum()}')
print(f'Duplicate percentage: {df.duplicated().sum()/len(df)*100:.1f}%')

Duplicate analysis:
Total records: 136757
Duplicates found: 70694
Duplicate percentage: 51.7%


In [4]:
# Remove duplicates
df = df.drop_duplicates()

In [5]:
print(f'After removing duplicates:')
print(f'Records remaining: {len(df)}')
print(f'Duplicates remaining: {df.duplicated().sum()}')

After removing duplicates:
Records remaining: 66063
Duplicates remaining: 0


**Key findings:**

Found 70,694 complete duplicate records (51.7% of dataset) where all sells in rows are identical. These are exact copies, not just similar salaries. All duplicate records have been removed, reducing dataset from 136,758 to approximately 66,000 unique records.

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Removing redundant columns</div>

Since we have salary_in_usd column which provides standardized salary values converted to USD, the original salary and salary_currency columns in various currencies are not informative for our analysis and can be removed.

In [6]:
df = df.drop(['salary', 'salary_currency'], axis=1)
print(f'Dataset shape after removing salary column: {df.shape}')

Dataset shape after removing salary column: (66063, 9)


<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Data types verification</div>

Before diving into the analysis, it's essential to ensure that all columns have the correct data types. This step helps prevent errors in calculations and visualizations and guarantees the accuracy of statistical operations.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66063 entries, 0 to 136756
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           66063 non-null  int64 
 1   experience_level    66063 non-null  object
 2   employment_type     66063 non-null  object
 3   job_title           66063 non-null  object
 4   salary_in_usd       66063 non-null  int64 
 5   employee_residence  66063 non-null  object
 6   remote_ratio        66063 non-null  int64 
 7   company_location    66063 non-null  object
 8   company_size        66063 non-null  object
dtypes: int64(3), object(6)
memory usage: 5.0+ MB


**Key findings:**

The dataset columns have correct data types, ensuring reliable analysis. All columns have complete data with no missing values (66,063 non-null entries for each column), which simplifies our cleaning process significantly.

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Descriptive statistics</div>

In [8]:
df.describe(include='all').T

count unique             top   freq           mean  \
work_year           66063.0    NaN             NaN    NaN    2024.374294   
experience_level      66063      4              SE  34743            NaN   
employment_type       66063      4              FT  65382            NaN   
job_title             66063    398  Data Scientist   6779            NaN   
salary_in_usd       66063.0    NaN             NaN    NaN  151148.848084   
employee_residence    66063    102              US  55377            NaN   
remote_ratio        66063.0    NaN             NaN    NaN      24.672661   
company_location      66063     95              US  55439            NaN   
company_size          66063      3               M  64319            NaN   

                             std      min      25%       50%       75%  \
work_year               0.730968   2020.0   2024.0    2024.0    2025.0   
experience_level             NaN      NaN      NaN       NaN       NaN   
employment_type              NaN      NaN      NaN       NaN       NaN   
job_title                    NaN      NaN      NaN       NaN       NaN   
salary_in_usd       77438.922472  15000.0  96000.0  138900.0  190067.5   
employee_residence           NaN      NaN      NaN       NaN       NaN   
remote_ratio           42.970711      0.0      0.0       0.0       0.0   
company_location             NaN      NaN      NaN       NaN       NaN   
company_size                 NaN      NaN      NaN       NaN       NaN   

                         max  
work_year             2025.0  
experience_level         NaN  
employment_type          NaN  
job_title                NaN  
salary_in_usd       800000.0  
employee_residence       NaN  
remote_ratio           100.0  
company_location         NaN  
company_size             NaN

**Key findings:**

The descriptive statistics align with the dataset description on Kaggle. All values appear normal and realistic for Data Science/AI/ML roles, with salary ranges, experience levels, and employment types matching expected patterns.

However, salary distribution raises concerns: maximum value of 800K USD compared to 75th percentile of 190K USD - nearly 4x difference. This indicates potential extreme outliers that require additional analysis and data cleaning.

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Extreme values analysis</div>

To identify potential salary outliers, I examine values above the 99,7th percentile. This threshold is commonly used in statistical analysis as it captures the top 1% of extreme values while maintaining a reasonable sample size for pattern analysis. Values above this threshold may represent either legitimate high-compensation roles or potential data quality issues.

In [9]:
percentile_99 = df['salary_in_usd'].quantile(0.99)
print(f'\n99th percentile salary threshold: ${percentile_99:,.0f}')
extreme_salaries = df[df['salary_in_usd'] > percentile_99]
print(f'\nNumber of extreme cases: {len(extreme_salaries)} ({len(extreme_salaries) / len(df) * 100:.2f}%)')


99th percentile salary threshold: $392,647

Number of extreme cases: 661 (1.00%)



I display the top 10 salaries above the 99th percentile to examine which roles, experience levels, and company sizes are associated with unusually high compensation values.

In [10]:
extreme_cases = (
    extreme_salaries[['job_title', 'experience_level', 'salary_in_usd', 'company_size', 'work_year', 'company_location']]
    .sort_values(by='salary_in_usd', ascending=False)
)
extreme_cases.head(10)

job_title experience_level  salary_in_usd  \
123027                AI Architect               MI         800000   
39563            Software Engineer               SE         800000   
16196                    Architect               SE         800000   
56674                Data Engineer               EN         793136   
123870                Data Analyst               EN         774000   
39889            Research Engineer               MI         750000   
37829            Software Engineer               SE         750000   
5247    Machine Learning Scientist               SE         750000   
127613               Data Engineer               MI         750000   
125256                Data Analyst               SE         750000   

       company_size  work_year company_location  
123027            M       2024               CA  
39563             M       2025               US  
16196             M       2025               US  
56674             M       2025               AT  
123870            M       2024               MX  
39889             M       2025               US  
37829             M       2025               US  
5247              M       2025               US  
127613            M       2023               US  
125256            M       2024               US

**Key findings:**

Analysis of salaries above 99th percentile (392,647 USD) reveals concerning patterns that suggest data quality issues:

- Too many "round" salary figures (800K USD, 750K USD)
- Entry-level positions with 793K USD salaries (unrealistic)  
- Disproportionate concentration in medium-sized companies (99% of extreme salaries)

These represent only 1% of the dataset. As the next step, I will address these extreme values to ensure data quality for reliable analysis.

<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Handling extreme outliers</div>

For proper handling of extreme salary values above the 99th percentile, I apply the following approach:

1. Identify the 99th percentile of salaries, marking the top 1% of values as potential outliers
2. Exclude these extreme values from the main dataset, creating a clean sample
3. Calculate maximum salaries by groups using features: job title, experience level, work year, company size, and country location based on the clean sample
4. In the original dataset, replace extreme values (above 99th percentile) with calculated group maximums from the clean sample
5. This approach preserves realistic upper salary boundaries while removing unrealistic outliers and maintaining adequate data distribution

In [11]:
# Fills target_col values with group maximum from group_cols.
def fill_by_group_max(df, target_col, group_cols):
    group_max = df.groupby(group_cols)[target_col].transform('max')
    return df[target_col].mask(df[target_col] > group_max, group_max)

In [12]:
# Calculate 99th percentile and create clean dataset
percentile_99 = df['salary_in_usd'].quantile(0.99)
clean_df = df[df['salary_in_usd'] <= percentile_99].copy()

# Define grouping columns
group_cols = ['job_title', 'experience_level', 'work_year', 'company_size', 'company_location']

# Calculate group maximums from clean data
group_max = clean_df.groupby(group_cols)['salary_in_usd'].max().reset_index()

# Merge group maximums with original dataset
df = df.merge(group_max.rename(columns={'salary_in_usd': 'group_max_salary'}), on=group_cols, how='left')

# Replace extreme values with group maximums
df['salary_in_usd'] = df.apply(
    lambda row: row['group_max_salary'] if row['salary_in_usd'] > percentile_99 else row['salary_in_usd'],
    axis=1
)

# Remove temporary column
df.drop(columns=['group_max_salary'], inplace=True)

In [13]:
# Check if any unresolved values remain in the salary column
print(f'Unresolved values in salary_in_usd: {df['salary_in_usd'].isna().sum()}')

Unresolved values in salary_in_usd: 7


As these 7 rows represent a negligible fraction of the dataset, and cannot be logically imputed, we will safely drop them without affecting the overall analysis

In [14]:
df = df.dropna(subset=['salary_in_usd'])

**Key findings:**

Successfully processed extreme salary outliers using group-based replacement. As these 7 rows represent a negligible fraction of the dataset, and cannot be logically imputed, we safely dropped them without affecting the overall analysis. The maximum salary is now capped at the 99th percentile, ensuring more realistic salary distribution.

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66056 entries, 0 to 66062
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           66056 non-null  int64  
 1   experience_level    66056 non-null  object 
 2   employment_type     66056 non-null  object 
 3   job_title           66056 non-null  object 
 4   salary_in_usd       66056 non-null  float64
 5   employee_residence  66056 non-null  object 
 6   remote_ratio        66056 non-null  int64  
 7   company_location    66056 non-null  object 
 8   company_size        66056 non-null  object 
dtypes: float64(1), int64(2), object(6)
memory usage: 5.0+ MB


<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Saving cleaned dataset</div>

Saving the cleaned and prepared dataset for further analysis in subsequent notebooks.

In [16]:
df.to_csv('cleaned_salaries.csv', index=False)
print('Cleaned dataset saved successfully!')

Cleaned dataset saved successfully!


<div style="color: #095AAD; font-weight: bold; font-size: 15px;">

### Data cleaning summary</div>

Through comprehensive data cleaning and preparation, I have successfully completed the following key steps:

**Work completed:**
- Loaded dataset containing 136,758 salary records with 11 variables  
- Removed 70,694 duplicate records (51.7% of original data)  
- Verified data types and confirmed their correctness  
- Removed redundant columns with original salary and currency information, keeping only the standardized salary_in_usd variable
- Identified and processed extreme salary outliers above the 99th percentile  
- Applied group-based outlier correction; removed 7 unresolved extreme cases due to lack of matching groups  

**Results achieved:**
- Clean dataset with 66,056 unique records ready for analysis  
- Realistic salary distribution with maximum capped at the 99th percentile  
- All critical data quality issues resolved while preserving structure and analytical value  

The dataset is now ready for comprehensive salary analysis.